File: backend/src/main/java/org/multimes/backend/core/tg/handler/TgHandler.java
package org.multimes.backend.core.tg.handler;

import com.pengrad.telegrambot.Callback;
import com.pengrad.telegrambot.TelegramBot;
import com.pengrad.telegrambot.model.Chat;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.request.GetUpdates;
import com.pengrad.telegrambot.request.SendMessage;
import com.pengrad.telegrambot.response.GetUpdatesResponse;

import org.multimes.backend.core.web.model.entities.Dialog;
import org.multimes.backend.core.web.model.entities.Message;
import org.multimes.backend.core.web.repository.interfaces.IDialogRepository;
import org.multimes.backend.core.web.repository.interfaces.IMessageRepository;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

@Component
public class TgHandler {
    private final TelegramBot bot;

    private final IMessageRepository messageRepository;
    private final IDialogRepository dialogRepository;

    public TgHandler(TelegramBot bot, IMessageRepository messageRepository, IDialogRepository dialogRepository) {
        this.bot = bot;
        this.messageRepository = messageRepository;
        this.dialogRepository = dialogRepository;
    }

    private int oldMesId = -1;
    private int offset = 0;

    private String getFullName(Chat chat) {
        String firstName = chat.firstName();
        String lastName = chat.lastName();
        StringBuffer usernameBuffer = new StringBuffer();
        if (firstName != null) {
            usernameBuffer.append(firstName);
        }
        if (lastName != null) {
            usernameBuffer.append(" ");
            usernameBuffer.append(lastName);
        }
        return usernameBuffer.toString();
    }

    private String getCurrentTime() {
        return LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm")).toString();
    }

    @Scheduled(fixedRate = 1000)
    public void newGetUpdates() {
        GetUpdates getUpdates = new GetUpdates().offset(offset);
        bot.execute(getUpdates, new Callback<GetUpdates, GetUpdatesResponse>() {
            @Override
            public void onResponse(GetUpdates request, GetUpdatesResponse response) {
                List<Update> updates = response.updates();
                if (updates != null) {
                    for (int i = 0; i < updates.size(); ++i) {
                        Update update = updates.get(i);
                        int mesId = update.message().messageId();
                        Chat chat = update.message().chat();
                        long chatId = chat.id();
                        String username = getFullName(chat);
                        int interId = dialogRepository.checkExistsWithIdInMessenger(chatId);
                        if (interId == -1) {
                            Dialog newDialog = new Dialog(-1, chatId, username, "telegram");
                            interId = dialogRepository.add(newDialog);
                        }
                        if (mesId != oldMesId) {
                            String text = update.message().text();
                            if (text == null || text.isEmpty()) {
                                text = "[UNSUPPORTED_FORMAT]";
                            }
                            String time = getCurrentTime();
                            messageRepository.add(new Message(-1, text, time, true, interId));
                            oldMesId = mesId;
                        }
                        if (i == updates.size() - 1) {
                            offset = update.updateId() + 1;
                        }
                    }
                }
            }

            @Override
            public void onFailure(GetUpdates request, IOException e) {
                System.out.println(e.getMessage());
            }
        });
    }

    public void sendMessage(long chatId, String message) {
        bot.execute(new SendMessage(chatId, message));
    }
}

File: backend/src/main/kotlin/org/multimes/backend/core/tg/handler/TgHandler.kt
package org.multimes.backend.core.tg.handler

import com.pengrad.telegrambot.Callback
import com.pengrad.telegrambot.TelegramBot
import com.pengrad.telegrambot.model.Chat
import com.pengrad.telegrambot.model.Update
import com.pengrad.telegrambot.request.GetUpdates
import com.pengrad.telegrambot.request.SendMessage
import com.pengrad.telegrambot.response.GetUpdatesResponse
import org.multimes.backend.core.web.model.entities.Dialog
import org.multimes.backend.core.web.model.entities.Message
import org.multimes.backend.core.web.repository.interfaces.IDialogRepository
import org.multimes.backend.core.web.repository.interfaces.IMessageRepository
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.io.IOException
import java.time.LocalTime
import java.time.format.DateTimeFormatter

@Component
class TgHandler(
    private val bot: TelegramBot,
    private val messageRepository: IMessageRepository,
    private val dialogRepository: IDialogRepository
) {

    private var oldMesId = -1
    private var offset = 0

    private fun getFullName(chat: Chat): String {
        val firstName = chat.firstName()
        val lastName = chat.lastName()
        return buildString {
            if (firstName != null) {
                append(firstName)
            }
            if (lastName != null) {
                append(" ")
                append(lastName)
            }
        }
    }

    private fun getCurrentTime(): String {
        return LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm"))
    }

    @Scheduled(fixedRate = 1000)
    fun newGetUpdates() {
        val getUpdates = GetUpdates().offset(offset)
        bot.execute(getUpdates, object : Callback<GetUpdates, GetUpdatesResponse> {
            override fun onResponse(request: GetUpdates, response: GetUpdatesResponse) {
                val updates = response.updates()
                if (updates != null) {
                    for (i in updates.indices) {
                        val update = updates[i]
                        val mesId = update.message().messageId()
                        val chat = update.message().chat()
                        val chatId = chat.id()
                        val username = getFullName(chat)
                        var interId = dialogRepository.checkExistsWithIdInMessenger(chatId)
                        if (interId == -1) {
                            val newDialog = Dialog(-1, chatId, username, "telegram")
                            interId = dialogRepository.add(newDialog)
                        }
                        if (mesId != oldMesId) {
                            val text = update.message().text()?.takeIf { it.isNotEmpty() } ?: "[UNSUPPORTED_FORMAT]"
                            val time = getCurrentTime()
                            messageRepository.add(Message(-1, text, time, true, interId))
                            oldMesId = mesId
                        }
                        if (i == updates.size - 1) {
                            offset = update.updateId() + 1
                        }
                    }
                }
            }

            override fun onFailure(request: GetUpdates, e: IOException) {
                println(e.message)
            }
        })
    }

    fun sendMessage(chatId: Long, message: String) {
        bot.execute(SendMessage(chatId, message))
    }
}
